internal/client/aws.go

```diff
@@ -0,0 +1,214 @@
+		InstanceIds: []string{vmName},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to start instance: %w", err)
+	}
+
+	return nil
+}
+
+func (a *AwsCli) StopInstance(ctx context.Context, vmName string) error {
+	_, err := a.client.StopInstances(ctx, &ec2.StopInstancesInput{
+		InstanceIds: []string{vmName},
+		// Forces the instances to stop. The instances do not have an opportunity to flush
+		// file system caches or file system metadata. If you use this option, you must
+		// perform file system check and repair procedures. This option is not recommended
+		// for Windows instances. Default: false
+		//Force:       force,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to stop instance: %w", err)
+	}
+
+	return nil
+}
+
+func (a *AwsCli) FindInstanceByTags(ctx context.Context, tags map[string]string) (*types.Instance, error) {
+	resp, err := a.client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		Filters: []types.Filter{
+			{
+				Name:   aws.String("tag:GARM_CONTROLLER_ID"),
+				Values: []string{tags["GARM_CONTROLLER_ID"]},
+			},
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []string{tags["Name"]},
+			},
+		},
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to find instances by tags: %w", err)
+	}
+
+	var instances []types.Instance
+	for _, reserv := range resp.Reservations {
+		instances = append(instances, reserv.Instances...)
+	}
+
+	return &instances[0], nil
+}
+
+// Describes the specified instances or all instances. If you specify instance
+// IDs, the output includes information for only the specified instances. If you
+// specify filters, the output includes information for only those instances that
+// meet the filter criteria.
+func (a *AwsCli) GetInstance(ctx context.Context, instance string) (*types.Instance, error) {
+	resp, err := a.client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []string{instance},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get instance: %w", err)
+	}
+
+	var instances []types.Instance
+	for _, reserv := range resp.Reservations {
+		instances = append(instances, reserv.Instances...)
+	}
+
+	if len(instances) == 0 {
+		return nil, fmt.Errorf("no such instance %s: %w", instance, errors.ErrNotFound)
+	}
+
+	return &instances[0], nil
+}
+
+// You can stop, start, and terminate EBS-backed instances. You can only terminate instance store-backed instances.
+// What happens to an instance differs if you stop it or terminate it. For example, when you stop an instance,
+// the root device and any other devices attached to the instance persist. When you terminate an instance,
+// any attached EBS volumes with the DeleteOnTermination block device mapping parameter set to true are
+// automatically deleted.
+func (a *AwsCli) TerminateInstance(ctx context.Context, vmName string) error {
+	_, err := a.client.TerminateInstances(ctx, &ec2.TerminateInstancesInput{
+		InstanceIds: []string{vmName},
+	})
+	if err != nil {
```

Comment:
Indeed the call will fail, updated it to check if the instance exists first:
```go
instance, _ := a.FindOneInstance(ctx, "", vmName)
	if instance.InstanceId == nil {
		return nil
	}
```

Reply:
If it fails with a specific error like `ErrNotFound` we can just check for that error and return nil. Between the `FindOneInstance()` call and the `Delete()` call, the instance can be removed, so this might still fail.

If, however, we check the error returned by the API, and can determine it's a `NotFound` error, we can just ignore it.